File: 0538-convert-bst-to-greater-tree/0538-convert-bst-to-greater-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
private:
    int sum;

public:
    TreeNode* recurse(TreeNode* root) {
        if (!root)
            return root;
        root->right = recurse(root->right);
        // 单层递归逻辑
        sum += root->val;
        root->val = sum;
        root->left = recurse(root->left);
        return root;
    }

    TreeNode* convertBST(TreeNode* root) {
        sum = 0;
        return recurse(root);
    }
};

File: 0538-convert-bst-to-greater-tree/0538-convert-bst-to-greater-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def __init__(self):
        self.sum = 0

    def convertBST(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        def recurse(root):
            if root is None:
                return root
            root.right = recurse(root.right)
            self.sum += root.val
            root.val = self.sum
            root.left = recurse(root.left)
            return root
        return recurse(root)